### ISSUE TITLE
Accessible vectors

### ISSUE THREAD

**Issue by danallsop** [ScottLogic/jumpstart-Leaflet#328]
> **Migrated from:** https://github.com/Leaflet/Leaflet/issues/7822
> **Original author:** danallsop
> **Original created:** 2021-11-28T17:19:11Z
---

**Is your feature request related to a problem? Please describe.**

> There is a larger issue that I really don't have a solution for: All vectors, not just markers, need to be tabbable. The way things are currently, only markers are tabbable. This means that VoiceOver and other accessibility technologies won't work for other vectors.

_Originally posted by danallsop in https://github.com/Leaflet/Leaflet/issues/3210#issuecomment-73268793_

**Comment by danallsop** [comment]
## Comments (4 total)

---
> **Comment by:** danallsop **on** 2022-02-04T12:48:53Z

Adding some substance to this issue:

For basic keyboard and screen reader accessibility, vectors would:

- [ ] Need to be focusable (i.e. set the SVG [`tabindex`](https://svgwg.org/svg2-draft/struct.html#tabindexattribute) attribute to `0`), similar to markers:

  https://github.com/Leaflet/Leaflet/blob/2960f58a02e0858f9403788f14fa9f4bd9fd2736/src/layer/marker/Marker.js#L226-L227

- [ ] Need a configurable [accessible name](https://www.w3.org/TR/accname/#dfn-accessible-name), e.g. via `aria-label` (similar to marker `alt`):

  https://github.com/Leaflet/Leaflet/blob/2960f58a02e0858f9403788f14fa9f4bd9fd2736/src/layer/marker/Marker.js#L43-L45

  And also tooltip via `<title>` (https://github.com/Leaflet/Leaflet/issues/4179) similar to marker `title`:

  https://github.com/Leaflet/Leaflet/blob/2960f58a02e0858f9403788f14fa9f4bd9fd2736/src/layer/marker/Marker.js#L39-L41

- [ ] If a vector binds to a popup or otherwise is clickable it should have `role="button"`, similar to markers: https://github.com/Leaflet/Leaflet/blob/2960f58a02e0858f9403788f14fa9f4bd9fd2736/src/layer/marker/Marker.js#L228

Further enhancements:

- [ ] If a vector binds to a popup it should convey the expanded/collapsed state of the popup using `aria-expanded="true|false"`, same as for markers: [#302](https://github.com/Leaflet/Leaflet/issues/7527). It should be noted that [`aria-expanded` relies on the presence of `role="button"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded#associated_roles).
---
> **Comment by:** danallsop **on** 2022-02-05T20:14:27Z

danallsop made a valid point in https://github.com/Leaflet/Leaflet/issues/4179#issuecomment-176726915:

> what if the line is rendered using a `L.Canvas` rendered instead? This functionality would be specific to the SVG renderer

Ideally (at least in the long term) the same keyboard functionality and accessible names would be provided independently of whether SVG or Canvas is used. That may be a bit of a challenge, and I haven't delved into canvas accessibility before.


<details>
<summary>List of potentially useful resources for reference</summary>

- https://webaim.org/blog/future-web-accessibility-html-canvas/:

  > All of the canvas accessibility solutions discussed in the previous section have one thing in common: they all require some nontrivial work on the part of the developer in order to make accessibility happen.
- https://www.w3.org/Talks/2014/0510-canvas-a11y/
- https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/hh968259(v=vs.85)
- https://www.oreilly.com/library/view/html5-canvas-2nd/9781449335847/ch01s12.html

  > To make the Canvas accessible, a method known as “Fallback DOM Concept,” or [sub dom](http://www.w3.org/wiki/Reading_text_in_canvas), was devised. Using this method, developers create a DOM element to match each element on the Canvas and put it in the sub dom.

- https://www.tpgi.com/html5-canvas-sub-dom/
- https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas#accessibility_concerns

The [Accessibility Object Model](https://github.com/WICG/aom/blob/gh-pages/explainer.md) (AOM) may be relevant in the (distant) future.

</details>

But maybe this feature (and https://github.com/Leaflet/Leaflet/issues/4179) should focus on the SVG use case only, to start with?

---
> **Comment by:** danallsop **on** 2022-04-02T11:56:30Z

For the Canvas use case (https://github.com/Leaflet/Leaflet/issues/7822#issuecomment-1030691886):

Azure Maps use [visually hidden elements](https://webaim.org/techniques/css/invisiblecontent/#offscreen) (they happen to use popups, but that's probably not desirable for Leaflet) to enable keyboard and screen reader accessibility for vectors in their [Accessible Shape Selection sample](https://samples.azuremaps.com/popups/accessible-shape-selection):

![azure-accessible-shapes](https://user-images.githubusercontent.com/26493779/161381915-515abfb6-f0e1-4d0a-abfc-5966abfdfb72.png)


---
> **Comment by:** danallsop **on** 2023-10-13T07:15:32Z

Hi danallsop,

Thanks for raising this. Valuable work! 👍🏻 

Linking to the PR again, for extra visibility: [#8251](https://github.com/Leaflet/Leaflet/issues/8251).
